### PR TITLE
llama: preserve field order in user-defined JSON schemas

### PR DESCRIPTION
**Pull request by bmizerany**
llama: preserve field order in user-defined JSON schemas

Previously we decoded and re-encoded JSON schemas during validation,
which served no purpose since json.RawMessage already validates JSON
syntax. Worse, the re-encoding lost field ordering from the original
schema, which affects inference quality during step-by-step reasoning.

While fixing this ordering issue by using json.RawMessage directly,
testing revealed that schema_to_grammar (from llama.cpp) also fails to
preserve field order during grammar generation. This appears to be the
root cause of inference degradation.

This change prevents us from mangling the user's original schema order,
but we still need to address the ordering issue in schema_to_grammar.
That will be a separate change.

Updates #7978

### CODE REVIEW

**Comment by iscy** [comment]
@bmizerany, llama.cpp's `json-schema-to-grammar` already ensures that the order is being preserved by using `nlohmann::ordered_json`. These PRs were opened in the past and kept the original key ordering while still relying on that facility:
#7588 
#6658

**Comment by bmizerany** [comment]
Skipping schema_to_grammer tests in prep for @ParthSareen's work on patching. Ready to merge.

File: llama/grammar_test.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	"additionalProperties": false
+      }
+    },
+    "final_answer": { "type": "string" }
+  },
+  "required": ["steps", "final_answer"],
+  "additionalProperties": false
+}`
+
+func TestIssue7978(t *testing.T) {
```
Comment (ParthSareen) [inline llama/grammar_test.go:32]:
```suggestion
func TestJSONSchemaOrdering(t *testing.T) {
```

File: llama/grammar_test.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+import (
+	"bufio"
+	"bytes"
+	"strings"
+	"testing"
+)
+
+// https://github.com/ollama/ollama/issues/7978
+const issue7978JSONSchema = `{
```
Comment (ParthSareen) [inline llama/grammar_test.go:11]:
```suggestion
const OrderedJSONSchema = `{
```
(or something along those lines)

File: llm/server.go [hunk 3]
```diff
@@ -744,22 +744,19 @@ func (s *llmServer) Completion(ctx context.Context, req CompletionRequest, fn fu
...
-			var schema llama.JsonSchema
-			err := json.Unmarshal(req.Format, &schema)
-			return schema, err
-		}(); err == nil {
-			request["grammar"] = schema.AsGrammar()
-		} else {
-			slog.Warn(`format is neither a schema or "json"`, "format", req.Format)
+			request["grammar"] = string(g)
+		default:
+			return errors.New(`invalid format: expected "json" or a JSON schema`)
```
Comment (ParthSareen) [inline llm/server.go:759]: I've considered letting this stay a warning - okay for now will figure behavior out in my PR

**Review by ParthSareen — APPROVED** [review]

File: llama/grammar_test.go (hunk 1 above)
Comment (bmizerany) [reply llama/grammar_test.go:32]: a) Thank you for using JSON, not Json
b) It's specifically about key ordering, not all things that can be order
c) This test is very specific to the issue the name refers to, so I'd like to keep it named as-is, later, when/if we have more generic ordering tests, we can call it `TestSchemaToJSONPreservedKeyOrder(...)` or something like that.

File: llama/grammar_test.go (hunk 2 above)
Comment (bmizerany) [reply llama/grammar_test.go:11]: See note above. This is specific to the issue. Keeping it as-is.

File: llm/server.go (hunk 3 above)
Comment (bmizerany) [reply llm/server.go:759]: It will lead to false positives. What does OpenAI do?

**Pull request merged**